clash-protocols-memmap/src/Protocols/MemoryMap.hs

```diff
@@ -0,0 +1,188 @@
+-- SPDX-FileCopyrightText: 2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FlexibleInstances #-}
+
+{-# OPTIONS_GHC -fconstraint-solver-iterations=10 #-}
+
+module Protocols.MemoryMap where
+
+import Clash.Prelude
+
+import Protocols
+import qualified Data.Map.Strict as Map
+import qualified Data.List as List
+import GHC.Stack (HasCallStack, SrcLoc)
+
+import Protocols.MemoryMap.FieldType (FieldType)
+
+-- | Absolute address or offset in bytes
+type Address = Integer
+
+-- | Size in bytes
+type Size = Integer
+
+-- | A protocol agnostic wrapper for memory mapped protocols. It provides a way
+-- for memory mapped subordinates to propagate their memory map to the top level,
+-- possibly with interconnects merging multiple memory maps.
+--
+-- The current implementation has a few objectives:
+--
+--   1. Provide a way for memory mapped peripherals to define their memory layout
+--
+--   2. Make it possible for designers to define hardware designs without specifying
+--      exact memory addresses if they can be mapped to arbitrary addresses, while
+--      not giving up the ability to.
+--
+--   3. Provide a way to check whether memory maps are valid, i.e. whether
+--      perhipherals do not overlap or exceed their allocated space.
+--
+--   4. Provide a way to generate human readable documentation
+--
+--   5. Provide a way to generate data structures for target languages. I.e., if
+--      a designer instantiates a memory mapped register on type @a@, it should
+--      be possible to generate an equivalent data structure in Rust or C.
+--
+
+data BackwardAnnotated (annotation :: Type) (a :: Type)
```

Comment:
`BackAnnotated`?
`BwdAnnotated`?

Reply:
It's a bit of a mouthful, yeah. We could also go with `BwdAnn` as the shortest option :sweat_smile:  Probably somewhere in between is good?

Reply:
"Back annotation" has a specific meaning in the hardware world: https://www.quora.com/What-is-meant-by-back-annotated-netlist-in-asic-design-flow. Just a heads up. I think `BwdAnn` is best (though I'm generally a bit meh on abbreviations like this).